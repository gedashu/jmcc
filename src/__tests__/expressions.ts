import JMCC from '../core';

const jmcc = new JMCC();

describe('выражения', () => {
  describe('математические', () => {
    test('мультипликативные', () => {
      const source = `
        var a = 2 * 2
        var b = 4 / 2
        var c = b % 3
        var chain = 1 * 2 / 3 % 4
      `;

      expect(jmcc.compileToJson(source)).toMatchSnapshot();
    });

    test('аддитивные', () => {
      const source = `
        var a = 2 + 2
        var b = a - 2
        var chain = 1 + 2 - 3
      `;

      expect(jmcc.compileToJson(source)).toMatchSnapshot();
    });

    test('со скобками', () => {
      const source = `
        var a = 2 + 2 * 2
        var b = (2 + 2) * 2
        var c = b / (1 + 3)
      `;

      expect(jmcc.compileToJson(source)).toMatchSnapshot();
    });

    test('с динамическими значениями', () => {
      const source = `
        var a = 1.clamp(0, 2)
        var b = a + 1
      `;

      expect(jmcc.compileToJson(source)).toMatchSnapshot();
    });
  });

  test('устанавливающие', () => {
    const source = `
      var a = 0
      a = 1
      a += 1
      a -= 1
      a *= 1
      a /= 1
      a %= 1
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('унарные', () => {
    const source = `
      var a = 0
      a++
      a--
      ++a
      --a
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });
});
