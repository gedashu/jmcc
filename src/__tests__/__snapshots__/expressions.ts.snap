// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`выражения математические аддитивные 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "4",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "%math(%var_local(a)-2)",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "chain",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "0",
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`выражения математические мультипликативные 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "4",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "2",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "c",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "%math(%var_local(b)%3)",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "chain",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "0.6666666666666666",
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`выражения математические с динамическими значениями 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_clamp",
          "values": Array [
            Object {
              "name": "min",
              "value": Object {
                "number": 0,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "%math(%var_local(a)+1)",
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`выражения математические со скобками 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "6",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "8",
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "c",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": "%math(%var_local(b)/4)",
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`выражения унарные 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 0,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_increment",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_increment",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_decrement",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_decrement",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`выражения устанавливающие 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 0,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_increment",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_decrement",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_multiply",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "scope": "local",
                    "type": "variable",
                    "variable": "a",
                  },
                  Object {
                    "number": 1,
                    "type": "number",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
          ],
        },
        Object {
          "action": "set_variable_divide",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "scope": "local",
                    "type": "variable",
                    "variable": "a",
                  },
                  Object {
                    "number": 1,
                    "type": "number",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "division_mode",
              "value": Object {
                "enum": "DEFAULT",
                "type": "enum",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_remainder",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "dividend",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "divisor",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;
